app/controllers/downloads_controller.rb

```diff
@@ -0,0 +1,98 @@
+class DownloadsController < ApplicationController
+  include Sufia::DownloadsControllerBehavior
+  
+  def authorize_download!
+    #authorize! :download, file
+  end
+  def send_file_contents
+        self.status = 200
+        prepare_file_headers
+        stream_body file.stream
+  end    
+  def show
+    if file.instance_of? ZipFile
+      return send_content
+    end
+    if file.new_record?
+      render_404
+    else
+      send_content
+    end
+  end
+  def stream_body(iostream)
+    iostream.each do |in_buff|
+      response.stream.write in_buff
+    end
+  ensure
+    response.stream.close
+  end
+  def send_content
+      response.headers['Accept-Ranges'] = 'bytes'
+
+      if request.head?
+        content_head
+      elsif request.headers['HTTP_RANGE']
+        send_range
+      else
+        print request.headers
+        send_file_contents
+        
+      end
+   end  
+  def load_file
+    file_path = params[:file]
+    file = GenericFile.find(params[:id]);
+    f = nil
+    #if file.instance_of? GenericFile and not file.resource_type.include? 'DataFile'
+    #  file = file.data_files.first
+    #end
+    
+    if !file_path && params[:datastream_id]
+      Deprecation.warn(DownloadBehavior, "Parameter `datastream_id` is deprecated and will be removed in hydra-head 10.0. Use `file` instead.")
+      file_path = params[:datastream_id]
+    end
+    if file.instance_of? DataFile
+      return super
+      #f = asset.attached_files[file_path] if file_path
+      #f ||= default_file 
+    end
+    
+    if file.instance_of? GenericFile
+        f = ZipFile.new(file.id << ".zip")
+        #Zip::OutputStream.open(f) { | z | } 
+        Zip::File.open(f.path, Zip::File::CREATE) do | zip |
```

Comment:
Is this thread-safe? What happens if two people try to download the same file at exactly the same time?